nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp

```diff
@@ -0,0 +1,186 @@
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Alexey Merzlyakov
+ *********************************************************************/
+
+#include "nav2_costmap_2d/costmap_filters/keepout_filter.hpp"
+
+namespace nav2_costmap_2d
+{
+
+KeepoutFilter::KeepoutFilter()
+: costmap_filter_info_sub_(nullptr), map_filter_sub_(nullptr), map_filter_(nullptr)
+{
+}
+
+void KeepoutFilter::initializeFilter(
+  const std::string costmap_filter_info_topic)
+{
+  // Setting new costmap filter info subscriber
+  RCLCPP_INFO(
+    node_->get_logger(),
+    "Setting a subscriber to %s costmap filter info",
+    costmap_filter_info_topic.c_str());
+  costmap_filter_info_sub_ = node_->create_subscription<nav2_msgs::msg::CostmapFilterInfo>(
+    costmap_filter_info_topic, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&KeepoutFilter::costmapFilterInfoCallback, this, std::placeholders::_1));
+
+  curr_time_ = node_->now();
+  prev_time_ = curr_time_;
+}
+
+void KeepoutFilter::costmapFilterInfoCallback(
+  const nav2_msgs::msg::CostmapFilterInfo::SharedPtr msg)
+{
+  // Resetting previous subscriber each time when new costmap filter information arrives
+  if (map_filter_sub_) {
+    RCLCPP_WARN(node_->get_logger(), "New map filter info arrived. Replacing old one.");
+    map_filter_sub_.reset();
+  }
+  map_filter_.reset();
+
+  // Setting new map filter subscriber
+  RCLCPP_INFO(
+    node_->get_logger(),
+    "Setting a subscriber to %s map filter",
+    msg->map_filter_topic.c_str());
+  map_filter_sub_ = node_->create_subscription<nav_msgs::msg::OccupancyGrid>(
+    msg->map_filter_topic, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&KeepoutFilter::mapFilterCallback, this, std::placeholders::_1));
+}
+
+void KeepoutFilter::mapFilterCallback(
+  const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  if (map_filter_) {
+    RCLCPP_WARN(node_->get_logger(), "New map filter arrived. Replacing old one.");
+    map_filter_.reset();
+  }
+  map_filter_ = std::make_unique<nav2_util::OccupancyGrid>(*msg);
+}
+
+void KeepoutFilter::process(
+  nav2_costmap_2d::Costmap2D & master_grid,
+  int min_i, int min_j, int max_i, int max_j,
+  double /*robot_x*/, double /*robot_y*/, double /*robot_yaw*/)
+{
+  if (!map_filter_) {
+    curr_time_ = node_->now();
+    // Show warning message every 2 seconds to not litter an output
+    if (curr_time_ - prev_time_ >= rclcpp::Duration(std::chrono::milliseconds(2000))) {
+      RCLCPP_WARN(node_->get_logger(), "Map filter was not received");
+      prev_time_ = curr_time_;
+    }
+    return;
+  }
+
+  unsigned int min_i_u, min_j_u, max_i_u, max_j_u;  // Unsigned versions of window bounds
+  double wx, wy;  // world coordinates
+
+  // unsigned<-signed conversions.
+  // All map bounds are positive, so this will be safe.
+  min_i_u = (unsigned int)min_i;
```

Comment:
Again, you must explicitly cast. `static_cast<unsigned int>(min_i)`

Reply:
All below as well

Reply:
Done starting from https://github.com/ros-planning/navigation2/commit/cfeeb1a0b96560bb68fa8f8119e27a62c65e7100#diff-f67a16397bb5979b7cbf4729ea53a451R172